tensorflow/python/keras/layers/preprocessing/preprocessing_stage_functional_test.py

```diff
@@ -0,0 +1,400 @@
+
+    l0 = PL()
+    y = l0(x)
+
+    l1 = PL()
+    z = l1(y)
+
+    stage = preprocessing_stage.FunctionalPreprocessingStage(
+        x, z
+    )
+    stage.compile()
+
+    # Test with NumPy array
+    one_array = np.ones((4, 3), dtype='float32')
+    stage.adapt(one_array)
+    self.assertEqual(l0.adapt_count, 1)
+    self.assertEqual(l1.adapt_count, 1)
+    self.assertLessEqual(l0.adapt_time, l1.adapt_time)
+
+    # Check call
+    z = stage(array_ops.ones((4, 3), dtype='float32'))
+    self.assertAllClose(z, np.ones((4, 3), dtype='float32') + 2.)
+
+    # Test with dataset
+    adapt_data = dataset_ops.Dataset.from_tensor_slices(one_array)
+    adapt_data = adapt_data.batch(2)  # 5 batches of 2 samples
+
+    stage.adapt(adapt_data)
+    self.assertEqual(l0.adapt_count, 2)
+    self.assertEqual(l1.adapt_count, 2)
+    self.assertLessEqual(l0.adapt_time, l1.adapt_time)
+
+    # Test error with bad data
+    with self.assertRaisesRegex(ValueError, 'requires a '):
+      stage.adapt(None)
+
+    # Disallow calling fit
+    with self.assertRaisesRegex(ValueError, 'Preprocessing stage'):
+      stage.fit(None)
+
+  def test_adapt_preprocessing_stage_with_list_input(self):
+
+    x0 = Input(shape=(3,))
+    x1 = Input(shape=(3,))
+    x2 = Input(shape=(3,))
+
+    l0 = PLMerge()
+    y = l0([x0, x1])
+
+    l1 = PLMerge()
+    y = l1([y, x2])
+
+    l2 = PLSplit()
+    z, y = l2(y)
+
+    stage = preprocessing_stage.FunctionalPreprocessingStage(
+        [x0, x1, x2], [y, z]
+    )
+    stage.compile()
+
+    # Test with NumPy array
+    one_array = np.ones((4, 3), dtype='float32')
+    stage.adapt([one_array, one_array, one_array])
+    self.assertEqual(l0.adapt_count, 1)
+    self.assertEqual(l1.adapt_count, 1)
+    self.assertEqual(l2.adapt_count, 1)
+    self.assertLessEqual(l0.adapt_time, l1.adapt_time)
+    self.assertLessEqual(l1.adapt_time, l2.adapt_time)
+
+    # Check call
+    y, z = stage([array_ops.ones((4, 3), dtype='float32'),
+                  array_ops.ones((4, 3), dtype='float32'),
+                  array_ops.ones((4, 3), dtype='float32')])
+    self.assertAllClose(y, np.ones((4, 3), dtype='float32') + 1.)
+    self.assertAllClose(z, np.ones((4, 3), dtype='float32') + 3.)
+
+    # Test with dataset
+    adapt_data = dataset_ops.Dataset.from_tensor_slices((one_array,
+                                                         one_array,
+                                                         one_array))
+    adapt_data = adapt_data.batch(2)  # 5 batches of 2 samples
+
+    stage.adapt(adapt_data)
+    self.assertEqual(l0.adapt_count, 2)
+    self.assertEqual(l1.adapt_count, 2)
+    self.assertEqual(l2.adapt_count, 2)
+    self.assertLessEqual(l0.adapt_time, l1.adapt_time)
+    self.assertLessEqual(l1.adapt_time, l2.adapt_time)
+
+    # Test error with bad data
+    with self.assertRaisesRegex(ValueError, 'requires a '):
+      stage.adapt(None)
+
+  def test_adapt_preprocessing_stage_with_dict_input(self):
```

Comment:
Is it possible to define a FunctionalPreprocessingStage with a dict of outputs?
Can we test this as well?

Reply:
Yes, I'll add the test case